dokka-integration-tests/gradle/projects/coroutines/coroutines.diff

```diff
@@ -1,72 +1,139 @@
 diff --git a/build.gradle b/build.gradle
-index e7d405e12..236310823 100644
+index e7d405e12..db5dcec66 100644
 --- a/build.gradle
 +++ b/build.gradle
-@@ -59,8 +59,9 @@ buildscript {
+@@ -52,17 +52,27 @@ buildscript {
      }
- 
+
+     repositories {
++        /* %{PROJECT_LOCAL_MAVEN_DIR}% */
+         mavenCentral()
+         maven { url "https://plugins.gradle.org/m2/" }
+         CommunityProjectsBuild.addDevRepositoryIfEnabled(delegate, project)
+-        mavenLocal()
++        //mavenLocal()
+     }
+
      dependencies {
-+        def dokkaVersion = System.getenv('DOKKA_VERSION')
          classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
 -        classpath "org.jetbrains.dokka:dokka-gradle-plugin:$dokka_version"
-+        classpath "org.jetbrains.dokka:dokka-gradle-plugin:$dokkaVersion"
++        classpath "org.jetbrains.dokka:dokka-gradle-plugin:${providers.gradleProperty("dokka_it_dokka_version").get()}"
          classpath "org.jetbrains.kotlinx:atomicfu-gradle-plugin:$atomicfu_version"
-         classpath "org.jetbrains.kotlinx:kotlinx-knit:$knit_version"
+-        classpath "org.jetbrains.kotlinx:kotlinx-knit:$knit_version"
++        classpath("org.jetbrains.kotlinx:kotlinx-knit:$knit_version") {
++            exclude(group: "org.jetbrains.kotlinx", module: "dokka-pathsaver-plugin")
++        }
++        classpath("org.jetbrains.kotlinx:dokka-pathsaver-plugin:$knit_version") {
++            exclude(group: "org.jetbrains.dokka", module: "templating-plugin")
++            exclude(group: "org.jetbrains.dokka", module: "dokka-base")
++        }
++        classpath("org.jetbrains.dokka:templating-plugin:${providers.gradleProperty("dokka_it_dokka_version").get()}")
++        classpath("org.jetbrains.dokka:dokka-base:${providers.gradleProperty("dokka_it_dokka_version").get()}")
```

Comment:
Knit uses Dokka v1.8.10 dependencies. Since the local dev-repo is marked as an exclusive source for `org.jetbrains.dokka` dependencies, and the repo doesn't contain 1.8.10 versions, the build will fail because it can't find the 1.8.10 dependencies.

Maybe there's a better way of doing this? Or update the version of Coroutines we're testing?

Reply:
> Knit uses Dokka v1.8.10 dependencies. Since the local dev-repo is marked as an exclusive source for org.jetbrains.dokka dependencies, and the repo doesn't contain 1.8.10 versions, the build will fail because it can't find the 1.8.10 dependencies.
>
> Maybe there's a better way of doing this? Or update the version of Coroutines we're testing?

Hmm, that's both good and bad that we found it, I didn't realize it would be a problem... 

One of the things that these tests test is the compatibility with Dokka plugins, such as knit. Especially for cases such as when AGP started shipping an older versions of Dokka, and it produced build errors if you had a newer version of Dokka in one of the subprojects (not in the root, which had older Dokka from AGP), and all sorts of things ended up on the classpath. So it would help to keep it working roughly the same way as it would work for the users, meaning coroutines will be running with an explicit new version of Dokka, and with knit that depends on the older version of Dokka.

> Or update the version of Coroutines we're testing?

Will it help? Coroutines use the published versions of knit that depend on published versions of Dokka, so if we update coroutines - knit would depend not on Dokka 1.8.10, but on Dokka 1.9.10, which doesn't seem to help us that much?

> Maybe there's a better way of doing this?

Just a thought: is there a reason for why the local dev repo should be the exclusive source for `org.jetbrains.dokka`? Can it be resolved with the order of the repositories? Meaning, use the local dev-repo as the first repository, so that it resolves all snapshot dependencies first, and then mavenCentral / mavenLocal after it so that it resolves other versions of Dokka and other dependencies if requested? I understand it will slow down the build (Gradle's lookup will be longer), but it's fine in this particular case, won't make much of a difference - integration tests are expected to be rather long and not really expected to be run locally often anyway.

Reply:
> > Maybe there's a better way of doing this?
> 
> Just a thought: is there a reason for why the local dev repo should be the exclusive source for `org.jetbrains.dokka`?

The exclusivity filter is there to ensure the integration tests only test the current Dokka version.

I wanted to prevent the scenario where 1. publishing to the local test-repo silently fails, and 2. the external projects pick up some other Dokka version (for example, from Maven Local).

If you think that risk is acceptable then we could rely on ordering, as you suggested.

Reply:
> I wanted to prevent the scenario where 1. publishing to the local test-repo silently fails, and 2. the external projects pick up some other Dokka version (for example, from Maven Local).

An idea: can the value of `dokka_it_dokka_version` be something specific to the integration tests? A long time ago [it was set](https://github.com/Kotlin/dokka/blob/1.8.20/integration-tests/gradle/projects/coroutines/coroutines.diff#L45) to `for-integration-tests-SNAPSHOT`, I don't know what for exactly, but I can assume now that it was done to avoid the problem you mentioned. The chances of having a `for-integration-tests-SNAPSHOT` Dokka version in Maven Local seem to be slim

It doesn't have to be `for-integration-tests-SNAPSHOT`, but something that is not `2.0.0-SNAPSHOT`, not a release, `-dev` or `-test` version, which could indeed be available elsewhere

Reply:
> An idea: can the value of `dokka_it_dokka_version` be something specific to the integration tests?

Unfortunately, it's a little more difficult than that :(

The version is set in multiple places (e.g. there's a generated `.properties` file in the plugin JAR that contains the version, the version is set into all of the Maven/Gradle metadata.) So in order to set it consistently, it needs to be set from the start. And since the integration-test project is 'just' a consumer of the Dokka subprojects, it's hard for it to influence the version.

_Maybe_ it could be possible if the whole Dokka project just set its version to 1.9.20-DEV by default, but _never_ publishing any -DEV version (we could add some checks in the Gradle tasks to prevent this). Or with some Composite Build hackery.


> I see that the tests are passing now - were you able to solve it in some other way?

Oh, the code I highlighted fixed the problem, so the code ran, but I was highlighting to ask if there was a better idea.